### PR TITLE
Fix for background download in fs cache

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix for background download in fs cache.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit b6a3a462f4c6a0bf9d04a3050c512c60a8d2a78b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55252/b6a3a462f4c6a0bf9d04a3050c512c60a8d2a78b/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6415392260">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55252/b6a3a462f4c6a0bf9d04a3050c512c60a8d2a78b/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/Cache/FileCache.cpp [hunk 1]
```diff
@@ -521,7 +521,7 @@ KeyMetadata::iterator FileCache::addFileSegment(
         result_state = state;
     }
 
-    auto file_segment = std::make_shared<FileSegment>(key, offset, size, result_state, settings, this, locked_key.getKeyMetadata());
+    auto file_segment = std::make_shared<FileSegment>(key, offset, size, result_state, settings, background_download_threads > 0, this, locked_key.getKeyMetadata());
```
Comment (novikd) [inline src/Interpreters/Cache/FileCache.cpp:524]: `0` usually means unlimited threads, but here it means we don't need to download a segment. Probably requires a comment somewhere.

File: src/Interpreters/Cache/Metadata.cpp [hunk 2]
```diff
@@ -536,13 +536,23 @@ void CacheMetadata::downloadThreadFunc()
...
+
+            downloadImpl(file_segment, memory);
         }
         catch (...)
         {
             if (holder)
             {
-                const auto & file_segment = holder->front();
+                auto & file_segment = holder->front();
+                file_segment.setDownloadFailed();
```
Comment (kssenii) [inline src/Interpreters/Cache/Metadata.cpp:554]: The main fix is in this line. It prevents submitting the file segment back to download queue in `~FileSegmentsHolder() -> FileSegment::complete()`

File: src/Interpreters/Cache/FileCache.cpp (hunk 1 above)
Comment (kssenii) [reply src/Interpreters/Cache/FileCache.cpp:524]: Added

**Review by novikd — APPROVED** [review]

**Comment by kssenii** [comment]
> Stress test (tsan) 
Upgrade check (tsan) 

https://github.com/ClickHouse/ClickHouse/pull/55260

**Comment by kssenii** [comment]
> Integration tests (asan) [4/6] — fail: 1

test_system_clusters_actual_information

> Stateless tests (release, DatabaseReplicated) [4/4] — fail: 1

00385_storage_file_and_clickhouse-local_app_long

> Upgrade check (msan) — check_status.tsv doesn't exists

https://github.com/ClickHouse/ClickHouse/issues/55306

**Pull request merged**